### PR TITLE
Feature: managed metadata

### PR DESCRIPTION
**Pull request by Zorin95670**

### CODE REVIEW

File: src/metadata/TerraformMetadata.js [hunk 1]
```diff
@@ -1,8 +1,207 @@
...
+      if (!validate(this.providers[provider])) {
+        errors.push({
+          provider,
+          errors: validate.errors,
+        });
+      }
+    });
+
+    if (errors.length > 0) {
+      throw new Error('Metadata is not valid', { cause: errors });
```
Comment (k-y0u) [inline src/metadata/TerraformMetadata.js:40]: `Metadata is not valid` -> `Metadata are not valid`
+
you check `providers` and finally it's `metadata` in the message... not pretty clear.

File: src/metadata/TerraformMetadata.js [hunk 2]
```diff
@@ -1,8 +1,207 @@
...
+
+    return {
+      components,
+      links,
+    };
+  }
+
+  /**
+   * Get link definitions from components.
+   * @param {TerraformComponentDefinition[]} components - Component definitions to parse.
```
Comment (k-y0u) [inline src/metadata/TerraformMetadata.js:62]: components, component definitions, not clear...

File: src/metadata/TerraformMetadata.js [hunk 3]
```diff
@@ -1,8 +1,207 @@
...
+    return links;
+  }
+
+  /**
+   * Set link definition from attribute in link definition array.
+   * @param {ComponentLinkDefinition[]} links - Link definition array.
+   * @param {String} componentType - Type of related component.
+   * @param {ComponentAttributeDefinition} definedAttribute - Attribute to parse.
+   */
+  getLinkDefinition(array, componentType, definedAttribute) {
```
Comment (k-y0u) [inline src/metadata/TerraformMetadata.js:80]: `getLinkDefinition` -> SetLinkDefinition

File: src/metadata/TerraformMetadata.js [hunk 4]
```diff
@@ -1,8 +1,207 @@
...
+      isContainer: provider.isContainer,
+    });
+
+    definition.parentTypes = this.getParentTypes(definition);
+
+    return definition;
+  }
+
+  /**
+   * Get block definition, can be module/data/resource/variable.
```
Comment (k-y0u) [inline src/metadata/TerraformMetadata.js:154]: `can be module/data/resource/variable.` repeated just under, so a bit useless

File: src/metadata/TerraformMetadata.js [hunk 5]
```diff
@@ -1,8 +1,207 @@
...
+  getAttributeDefinition(attribute) {
+    const subAttributes = attribute.attributes || [];
+    return new ComponentAttributeDefinition({
+      ...attribute,
+      definedAttributes: subAttributes.map(this.getAttributeDefinition),
+    });
+  }
+
+  /**
+   * Get all possible parent types container.
```
Comment (k-y0u) [inline src/metadata/TerraformMetadata.js:190]: `types container` -> `container types`

File: src/metadata/TerraformMetadata.js [hunk 6]
```diff
@@ -1,8 +1,207 @@
...
+   * @param {TerraformComponentDefinition} componentDefinition - Definition to get all parent types.
+   * @returns {String[]} All possible parent types container.
```
Comment (k-y0u) [inline src/metadata/TerraformMetadata.js:192]: same here

File: tests/unit/metadata/TerraformMetadata.spec.js [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+            icon: 'variable_icon',
+            model: 'variable_model',
+            definedAttributes: [],
+          }),
+        ]);
+      });
+    });
+
+    describe('Test method: getProviderDefinition', () => {
+
```
Comment (k-y0u) [inline tests/unit/metadata/TerraformMetadata.spec.js:127]: hello empty !

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: src/models/TerraformComponentDefinition.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import { ComponentDefinition } from 'leto-modelizer-plugin-core';
+
+/**
+ * Specific Terraform component definition.
+ * @extends {ComponentDefinition}
+ */
+class TerraformComponentDefinition extends ComponentDefinition {
+  /**
+   * Override Component Definition construction with blockType and provider properties.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/TerraformComponentDefinition.js:9]: done

File: src/models/TerraformComponentDefinition.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+   * @param {String} [props.blockType] - Type of Terraform block.
+   * @param {String} [props.provider] - Related provider of the terraform block.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/TerraformComponentDefinition.js:11]: done

File: tests/unit/models/TerraformComponentDefinition.spec.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import TerraformComponentDefinition from 'src/models/TerraformComponentDefinition';
+
+describe('Test class: TerraformComponentDefinition', () => {
+  describe('Test constructor', () => {
+    it('Check variable instantiation', () => {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/models/TerraformComponentDefinition.spec.js:5]: done

File: src/metadata/TerraformMetadata.js (hunk 1 above)
Comment (Zorin95670) [reply src/metadata/TerraformMetadata.js:40]: because it's the provider metadata :)

File: src/metadata/TerraformMetadata.js (hunk 2 above)
Comment (Zorin95670) [reply src/metadata/TerraformMetadata.js:62]: done

File: src/metadata/TerraformMetadata.js (hunk 3 above)
Comment (Zorin95670) [reply src/metadata/TerraformMetadata.js:80]: done

File: src/metadata/TerraformMetadata.js (hunk 4 above)
Comment (Zorin95670) [reply src/metadata/TerraformMetadata.js:154]: done

File: src/metadata/TerraformMetadata.js (hunk 5 above)
Comment (Zorin95670) [reply src/metadata/TerraformMetadata.js:190]: done

File: src/metadata/TerraformMetadata.js (hunk 6 above)
Comment (Zorin95670) [reply src/metadata/TerraformMetadata.js:192]: done

File: tests/unit/metadata/TerraformMetadata.spec.js (hunk 7 above)
Comment (Zorin95670) [reply tests/unit/metadata/TerraformMetadata.spec.js:127]: oups, done

**Review by k-y0u — APPROVED** [review]

**Pull request merged**